internal/proxy/shard_client.go

```diff
@@ -176,90 +182,78 @@ func newShardClientMgr(options ...shardClientMgrOpt) *shardClientMgrImpl {
 	c.clientCreator = creator
 }
 
-// Warning this method may modify parameter `oldLeaders`
-func (c *shardClientMgrImpl) UpdateShardLeaders(oldLeaders map[string][]nodeInfo, newLeaders map[string][]nodeInfo) error {
-	oldLocalMap := make(map[UniqueID]*nodeInfo)
-	for _, nodes := range oldLeaders {
-		for i := range nodes {
-			n := &nodes[i]
-			_, ok := oldLocalMap[n.nodeID]
-			if !ok {
-				oldLocalMap[n.nodeID] = n
-			}
-		}
-	}
-	newLocalMap := make(map[UniqueID]*nodeInfo)
-
-	for _, nodes := range newLeaders {
-		for i := range nodes {
-			n := &nodes[i]
-			_, ok := oldLocalMap[n.nodeID]
-			if !ok {
-				_, ok2 := newLocalMap[n.nodeID]
-				if !ok2 {
-					newLocalMap[n.nodeID] = n
-				}
-			}
-			delete(oldLocalMap, n.nodeID)
-		}
-	}
-	c.clients.Lock()
-	defer c.clients.Unlock()
+func (c *shardClientMgrImpl) GetClient(ctx context.Context, info nodeInfo) (types.QueryNodeClient, error) {
+	c.clients.RLock()
+	client, ok := c.clients.data[info.nodeID]
+	c.clients.RUnlock()
 
-	for _, node := range newLocalMap {
-		client, ok := c.clients.data[node.nodeID]
-		if ok {
-			client.inc()
-		} else {
-			// context.Background() is useless
-			// TODO QueryNode NewClient remove ctx parameter
-			// TODO Remove Init && Start interface in QueryNode client
-			if c.clientCreator == nil {
-				return fmt.Errorf("clientCreator function is nil")
-			}
-			client, err := newPoolingShardClient(node, c.clientCreator)
+	if !ok {
+		c.clients.Lock()
+		// Check again after acquiring the lock
+		client, ok = c.clients.data[info.nodeID]
+		if !ok {
+			// Create a new client if it doesn't exist
+			newClient, err := newShardClient(info, c.clientCreator)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			c.clients.data[node.nodeID] = client
+			c.clients.data[info.nodeID] = newClient
+			client = newClient
 		}
+		c.clients.Unlock()
 	}
-	for _, node := range oldLocalMap {
-		client, ok := c.clients.data[node.nodeID]
-		if ok && client.dec() {
-			delete(c.clients.data, node.nodeID)
+
+	return client.getClient(ctx)
+}
+
+func (c *shardClientMgrImpl) PurgeClient() {
+	ticker := time.NewTicker(600 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-c.closeCh:
+			return
+		case <-ticker.C:
+			shardLocations := globalMetaCache.ListShardLocation()
+			for nodeID := range c.clients.data {
+				if _, ok := shardLocations[nodeID]; !ok {
+					c.clients.Lock()
+					delete(c.clients.data, nodeID)
+					c.clients.Unlock()
+					c.ReleaseClient(nodeID)
+				}
+			}
 		}
 	}
-	return nil
 }
 
-func (c *shardClientMgrImpl) GetClient(ctx context.Context, nodeID UniqueID) (types.QueryNodeClient, error) {
-	c.clients.RLock()
-	client, ok := c.clients.data[nodeID]
-	c.clients.RUnlock()
-
-	if !ok {
-		return nil, fmt.Errorf("can not find client of node %d", nodeID)
+func (c *shardClientMgrImpl) ReleaseClient(nodeID int64) {
+	c.clients.Lock()
+	defer c.clients.Unlock()
+	if client, ok := c.clients.data[nodeID]; ok {
+		client.Release()
```

Comment:
Why does the client object remain in c.clients.data even after calling client.Release() ?